casper/src/main/scala/coop/rchain/casper/CasperConf.scala

```diff
@@ -1,17 +1,88 @@
 package coop.rchain.casper
 
+import cats.Monad
+import cats.implicits._
+
+import com.google.protobuf.ByteString
+
+import coop.rchain.catscontrib.Capture
 import coop.rchain.crypto.codec.Base16
+import coop.rchain.crypto.signatures.Ed25519
+import coop.rchain.shared.{Log, LogSource}
 
 import java.nio.file.Path
 
+import scala.io.Source
+import scala.util.{Failure, Success, Try}
+
 case class CasperConf(
     publicKeyBase16: Option[String],
     privateKeyBase16: Option[String],
     sigAlgorithm: String,
     bondsFile: Option[String],
+    knownValidatorsFile: Option[String],
     numValidators: Int,
-    validatorsPath: Path
+    validatorsPath: Path,
+    createGenesis: Boolean
 ) {
   val publicKey: Option[Array[Byte]]  = publicKeyBase16.map(Base16.decode)
   val privateKey: Option[Array[Byte]] = privateKeyBase16.map(Base16.decode)
 }
+
+object CasperConf {
+  private implicit val logSource: LogSource = LogSource(this.getClass)
+
+  def parseValidatorsFile[F[_]: Monad: Capture: Log](
+      knownValidatorsFile: Option[String]): F[Set[ByteString]] =
+    knownValidatorsFile match {
+      //TODO: Add default set? Throw error?
+      case None => Set.empty[ByteString].pure[F]
+
+      case Some(file) =>
+        Capture[F]
+          .capture {
+            Try(
+              Source
+                .fromFile(file)
+                .getLines()
+                .map(line => ByteString.copyFrom(Base16.decode(line)))
+                .toSet)
+          }
+          .flatMap {
+            case Success(validators) => validators.pure[F]
+
+            case Failure(ex) =>
+              Log[F]
+                .error(s"Error while parsing known validators file; $ex: ${ex.getMessage}")
+                .map[Set[ByteString]](_ => throw ex)
+          }
+    }
+
+  def publicKey(givenPublicKey: Option[Array[Byte]],
+                sigAlgorithm: String,
+                privateKey: Array[Byte]): Array[Byte] = {
+
+    val maybeInferred = sigAlgorithm match {
+      case "ed25519" =>
+        Try(Ed25519.toPublic(privateKey)).toOption
+
+      case _ => None
+    }
+
+    (maybeInferred, givenPublicKey) match {
+      case (Some(k1), Some(k2)) =>
+        if (keysMatch(k1, k2)) k1
+        else throw new Exception("Public key not compatible with given private key!")
+
+      case (Some(k1), None) => k1
+
+      case (None, Some(k2)) => k2
```

Comment:
Hmm if we failed to infer the Ed25519 public key from the private key, that is a red flag. I understand with other algorithms this might be ok but I think we at least should add a `// TODO: Think again` comment.